src/i2p.cpp

```diff
@@ -0,0 +1,432 @@
+                accepted.socket = sock.Release();
+
+                err_wait = err_wait_begin;
+
+                return true;
+            }
+        } catch (const std::runtime_error& e) {
+            Log("Error accepting: %s", e.what());
+            m_interrupt->sleep_for(err_wait);
+            if (err_wait < err_wait_cap) {
+                err_wait *= 2;
+            }
+        }
+    }
+    return false;
+}
+
+bool Session::Connect(const CService& to, Connection& conn, bool& proxy_error)
+{
+    proxy_error = true;
+
+    std::string session_id;
+    Sock sock;
+    conn.peer = to;
+
+    try {
+        {
+            LOCK(m_mutex);
+            CreateIfNotCreatedAlready();
+            session_id = m_session_id;
+            conn.me = m_my_addr;
+            sock = Hello();
+        }
+
+        const Reply& lookup_reply =
+            SendRequestAndGetReply(sock, strprintf("NAMING LOOKUP NAME=%s", to.ToStringIP()));
+
+        const std::string& dest = lookup_reply.Get("VALUE");
+
+        const Reply& connect_reply = SendRequestAndGetReply(
+            sock, strprintf("STREAM CONNECT ID=%s DESTINATION=%s SILENT=false", session_id, dest),
+            false);
+
+        const std::string& result = connect_reply.Get("RESULT");
+
+        if (result == "OK") {
+            conn.socket = sock.Release();
+            return true;
+        }
+
+        if (result == "INVALID_ID") {
+            LOCK(m_mutex);
+            Disconnect();
+            throw std::runtime_error("Invalid session id");
+        }
+
+        if (result == "CANT_REACH_PEER" || result == "TIMEOUT") {
+            proxy_error = false;
+        }
+
+        throw std::runtime_error(strprintf("'%s'", connect_reply.full));
+    } catch (const std::runtime_error& e) {
+        Log("Error connecting to %s: %s", to.ToString(), e.what());
+        return false;
+    }
+}
+
+// Private methods
+
+std::string Session::Reply::Get(const std::string& key) const
+{
+    const auto& pos = keys.find(key);
+    if (pos == keys.end() || !pos->second.has_value()) {
+        throw std::runtime_error(
+            strprintf("Missing %s= in the reply to '%s': '%s'", key, request, full));
+    }
+    return pos->second.value();
+}
+
+template <typename... Args>
+void Session::Log(const std::string& fmt, const Args&... args) const
+{
+    LogPrint(BCLog::I2P, "I2P: %s\n", tfm::format(fmt, args...));
+}
+
+Session::Reply Session::SendRequestAndGetReply(const Sock& sock,
+                                               const std::string& request,
+                                               bool check_result_ok) const
+{
+    SendComplete(sock.Get(), request + "\n", MAX_WAIT_FOR_IO, *m_interrupt);
+
+    Reply reply;
+
+
+    // Don't log the full "SESSION CREATE ..." because it contains our private key.
+    reply.request = request.substr(0, 14) == "SESSION CREATE" ? "SESSION CREATE ..." : request;
```

Comment:
If there is some kind of `startsWith` function i think it would be better.

Reply:
[`HasPrefix()`](https://github.com/bitcoin/bitcoin/blob/30e664dcce1a9adb9ba9a29e4f0cf809767870dd/src/util/string.h#L82-L88) is a good candidate, would need some enchanting.